assets/js/base/context/cart-checkout/payment-methods/payment-method-data-context.js

```diff
@@ -341,6 +341,11 @@ export const PaymentMethodDataProvider = ( { children } ) => {
 			return;
 		}
 
+		// Customer has saved card/payment methods - no need to set a default.
+		if ( customerPaymentMethods ) {
+			return;
+		}
+
 		// If there's no active payment method, or the active payment method has
 		// been removed (e.g. COD vs shipping methods), set one as active.
 		if (
```

Comment:
There is an alternative approach that doesn't require the `customerPaymentMethods` dependency. It would involve using `useState` functional updates. Something like:

```JSX
setActive( ( currentActivePaymentMethod ) => {
	// If there's no active payment method, or the active payment method has
	// been removed (e.g. COD vs shipping methods), set one as active.
	if (
		! currentActivePaymentMethod ||
		! paymentMethodKeys.includes( currentActivePaymentMethod )
	) {
		return Object.keys( paymentData.paymentMethods )[ 0 ];
	}
	return currentActivePaymentMethod;
} );
```

This would ensure `currentActivePaymentMethod` is evaluated right at the moment when the state is updated, so the payment method activated in `SavedPaymentMethodOptions` would already be active.

I didn't do a lot of testing of this approach, so feel free to ignore, but wanted to propose it in case you think it might be a good alternative.

Reply:
Smart! I like this, and of course since it's more general it avoids the issue with my truthy test below. Thanks for picking up on this, much better fix 🚀 

The crux of the issue is that we were previously using a stale copy of `currentActivePaymentMethod`; so it makes sense that we'd fix this by using the functional form of `setState`. The principle to keep in mind here (just reminding myself!) - [if a state update depends on the previous value, use functional form](https://reactjs.org/docs/faq-state.html#how-do-i-update-state-with-values-that-depend-on-the-current-state). 

I've made the change and I'll do some more testing of a few scenarios. Note the other change from the sketch above is the reset of checkout status to `PRISTINE` when applying the default.